cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py

```diff
@@ -0,0 +1,227 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Utility methods for decomposing arbitrary n-qubit (2^n x 2^n) unitary.
+
+Based on:
+Synthesis of Quantum Logic Circuits. Tech. rep. 2006,
+https://arxiv.org/abs/quant-ph/0406176
+"""
+from typing import List, Callable, TYPE_CHECKING
+
+from scipy.linalg import cossin
+
+import numpy as np
+
+from cirq import ops
+from cirq.linalg import decompositions, predicates
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def quantum_shannon_decomposition(qubits: 'List[cirq.Qid]', u: np.ndarray) -> 'cirq.OP_TREE':
+    """Decomposes n-qubit unitary into CX/YPow/ZPow/CNOT gates, preserving global phase.
+
+    The algorithm is described in Shende et al.:
+    Synthesis of Quantum Logic Circuits. Tech. rep. 2006,
+    https://arxiv.org/abs/quant-ph/0406176
+
+    Args:
+        qubits: List of qubits in order of significance
+        u: Numpy array for unitary matrix representing gate to be decomposed
+
+    Calls:
+        (Base Case)
+        1. _single_qubit_decomposition
+            OR
+        (Recursive Case)
+        1. _msb_demuxer
+        2. _multiplexed_cossin
+        3. _msb_demuxer
+
+    Yields:
+        A single 2-qubit or 1-qubit operations from OP TREE
+        composed from the set
+           { CNOT, rz, ry, ZPowGate }
+
+    Raises:
+        ValueError: If the u matrix is non-unitary
+        ValueError: If the u matrix is not of shape (2^n,2^n)
+    """
+    if not predicates.is_unitary(u):  # Check that u is unitary
+        raise ValueError(
+            "Expected input matrix u to be unitary, \
+                but it fails cirq.is_unitary check"
+        )
+
+    n = u.shape[0]
+    if n & (n - 1):
+        raise ValueError(
+            f"Expected input matrix u to be a (2^n x 2^n) shaped numpy array, \
+                but instead got shape {u.shape}"
+        )
+
+    if n == 2:
+        # Yield a single-qubit decomp if u is 2x2
+        yield from _single_qubit_decomposition(qubits[0], u)
+        return
+
+    # Perform a cosine-sine (linalg) decomposition on u
+    #   X   =   [ u1 , 0  ] [ cos(theta) , -sin(theta) ] [ v1 , 0  ]
+    #           [ 0  , u2 ] [ sin(theta) ,  cos(theta) ] [ 0  , v2 ]
+    (u1, u2), theta, (v1, v2) = cossin(u, n / 2, n / 2, separate=True)
+
+    # Yield ops from decomposition of multiplexed v1/v2 part
+    yield from _msb_demuxer(qubits, v1, v2)
+
+    # Observe that middle part looks like Σ_i( Ry(theta_i)⊗|i><i| )
+    # Then most significant qubit is Ry multiplexed over all other qubits
+    # Yield ops from multiplexed Ry part
+    yield from _multiplexed_cossin(qubits, theta, ops.ry)
+
+    # Yield ops from decomposition of multiplexed u1/u2 part
+    yield from _msb_demuxer(qubits, u1, u2)
+
+
+def _single_qubit_decomposition(qubit: 'cirq.Qid', u: np.ndarray) -> 'cirq.OP_TREE':
```

Comment:
As discussed offline, consider updating the return type if internal helpers to `cirq.ops.OpTree` instead, which should be enough to fix errors in `yield from ()` statements.